show pull progress in terminal for inflight pull requests

samsabed opened the pull request:
Show pull progress when another client is also performing a same image pull request
This PR does it only on the pulled image level, and not layers, push
fixes #8385

Signed-off-by: Sam Abed sam.abed@gmail.com

cpuguy83 commented:
FYI this was attempted before in #10683 but I'd found some issues that couldn't be fixed with that particular implementation (which also relied on jobs, which are now gone). #10683 design was OK'd by @dmcgowan

samsabed commented:
@cpuguy83 thanks for the tip, this PR passes the python tests.

dmcgowan commented on graph/pull.go:
```diff
@@ -44,17 +44,20 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
-			<-c
+			ps.AddObserver(imagePullConfig.OutStream)
+			<-ps.c
+			imagePullConfig.OutStream.Write(sf.FormatStatus("", "Repository %s Downloaded.", repoInfo.LocalName))
```
This should probably be formatted like the output from "WriteStatus"

dmcgowan commented:
The code looks good, but I will need to spend some time this week testing it out. Its hard for the integration tests to pick up on some of the possible odd behavior with the output streams.

Also looks like you need to squash your "SQUASH" commit.

samsabed commented:
@dmcgowan yeah just thought to keep the noise down, let me know if the change is something that is of interest and I'll do a rebase / fixups

icecrime commented:
@samsabed Thanks for the contribution! Indeed it's something of interest, I think we can move forward with the code review if you can please rebase and squash your commits.

Commit:
show pull progress in terminal for inflight pull requests

fixes #8385

Signed-off-by: Sam Abed <sam.abed@gmail.com>

samsabed commented:
@icecrime great, I've rebased the changes, and checked against v2, I'll need to get a v1 registry setup to test

dmcgowan commented:
@samsabed the progress bar no longer shows up, it only shows the number of bytes pulled

dmcgowan commented:
Looks like might work in v2 and have issues with v1, possibly the fallback messes it up
v1 pull (after fallback)
https://asciinema.org/a/4fsr7slbccnmwrnormoemgie9

v2 pull
https://asciinema.org/a/bfmdm7dzpj8m7pgqpncrtsri6

samsabed commented:
@dmcgowan You mean how it's a counter when pulling from V1 ? That's how it looks on master, I've reset/clean my clone of master to confirm. BTW That asciinema is impressive ever time I see it

dmcgowan commented:
@samsabed confirmed, it does appear broken in master as well

dmcgowan commented on graph/pull.go:
```diff
@@ -43,17 +43,19 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
+	ps, err := s.poolAdd("pull", utils.ImageReference(repoInfo.LocalName, tag))
 	if err != nil {
-		if c != nil {
+		if ps.c != nil {
```
Could we add a function to check and wait on c? This way ProgressStatus could isn't tied to the graph package.

dmcgowan commented:
One suggestion on the exposed interface to ProgressStatus

However from testing it looks good, very nice work :smile_cat: 
Lets address those comments but either way LGTM

samsabed commented:
@dmcgowan Thanks for the review, makes sense I've moved the class under pkg/progressreader/ not sure if it's the right place.
Added Wait() to encapsulate the blocking
Renamed Finish to Done()

dmcgowan commented on graph/load.go:
```diff
@@ -109,13 +109,12 @@ func (s *TagStore) recursiveLoad(address, tmpImageDir string) error {
+		if ps, err := s.poolAdd("pull", "layer:"+img.ID); err != nil {
+			werr := ps.Wait(nil, nil)
+			if werr == nil {
 				logrus.Debugf("Image (id: %s) load is already running, waiting: %v", img.ID, err)
```
This log message now occurs after the wait

dmcgowan commented:
Changes still look good, just address the debug message

dmcgowan commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,71 @@
+	ps.Unlock()
+
+	if channel == nil {
+		// super defensive, not user facing message, channel was not allocated
```
If this isn't user facing, and this error appears to be swallowed, why throw an error at all?

Is the expectation that channel would be nil after Done is called? I don't see `pc.c` getting set to nil.

dmcgowan replied:
To clarify, my expectation would be `ps.c` getting set to nil when `Done()` is called, and these errors would at least be logged at Debug level when returned from Wait.

samsabed commented:
@dmcgowan I had kept the previous pattern of checking if the channel was nil as I wasn't 100% sure if it protecting against some GO gc/race

I've changed it to not check for a nil channel in the graph/pull and just logging a Debug in Wait() as you've suggested. The channel shouldn't be nil 'cause it's not being set to nil and I don't think there is gc or race issue here.
for the case of a dev "miss-using" poolAdd e.g. 
`poolAdd("not_a_pull_or_push", "img:"+img.ID)`
then that will just hard fail with a nil pointer error. what do you think ?

I can change the poolAdd/remove to pullPoolAdd/pushPoolAdd in another PR if you haven't suffered reviewer fatigue : )
Thanks again for the review.

dmcgowan commented:
LGTM

If you wanted to have a follow up PR, I would ask for the ability for Wait to return an error and being able to set an error on done. In this case if an error occurred on the pull, the observers would also return the error and set a non-zero status.

samsabed commented:
@dmcgowan Thanks and will follow up with a PR to address your recommendation  
Is it now pending a merge or needs another reviewer , just don't want to do a rebase :)

cc/  @icecrime

dmcgowan commented:
Ping @stevvooe @RichardScothern

stevvooe commented on graph/load.go:
```diff
@@ -109,14 +109,10 @@ func (s *TagStore) recursiveLoad(address, tmpImageDir string) error {
-			return err
+		if ps, err := s.poolAdd("pull", "layer:"+img.ID); err != nil {
+			logrus.Debugf("Image (id: %s) load is already running, waiting: %v", img.ID, err)
+			ps.Wait(nil, nil)
```
`nil` arguments make me nervous here. Why aren't we passing anything in?

stevvooe commented on graph/pull.go:
```diff
@@ -265,28 +263,26 @@ func (s *TagStore) pullRepository(r *registry.Session, out io.Writer, repoInfo *
 			defer s.poolRemove("pull", "img:"+img.ID)
 
-			out.Write(sf.FormatProgress(stringid.TruncateID(img.ID), fmt.Sprintf("Pulling image (%s) from %s", img.Tag, repoInfo.CanonicalName), nil))
+			ps.Write(sf.FormatProgress(stringid.TruncateID(img.ID), fmt.Sprintf("Pulling image (%s) from %s", img.Tag, repoInfo.CanonicalName), nil))
```
Please check your errors.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/docker/docker/vendor/src/github.com/Sirupsen/logrus"
+)
+
+type ProgressStatus struct {
```
Please document all exported types according to [Go's Documentation Guidelines](http://blog.golang.org/godoc-documenting-go-code). We cannot be having a type like this without a single comment about what it is supposed to do.

aaronlehmann replied:
Documented on my branch

stevvooe commented:
The goals of this PR are spot on, but I'm worried we might be headed further down the rabbit hole. Indeed, we'd like to ensure that downloads cannot are unique but this PR mixes this with progress reporting and status. It also relies on correctly calling it several places, rather than ensuring that the scope of a given pull ensures a lock is taking. This status + locking side-effecting makes me nervous for future changes.

That is not to say we shouldn't consider merging this, but we really need an abstraction to correctly handle this. It is very hard to verify and get right through inspection alone.

samsabed commented:
> The goals of this PR are spot on, but I'm worried we might be headed further down the rabbit hole. Indeed, we'd like to ensure that downloads cannot are unique but this PR mixes this with progress reporting and status. 

They need to be mixed a bit so that we can report on the progress of another client's download

> It also relies on correctly calling it several places, rather than ensuring that the scope of a given pull ensures a lock is taking. 

The branching around `s.poolAdd` ? That's not something this PR changed. can you provide an example of what you mean

> This status + locking side-effecting makes me nervous for future changes.

`ps.Wait` ? that encapsulates updating the current client progress bar and waiting for the other pull to finish. I don't see a locking side effect, that's it's intended purpose, breaking it up, will add boilerplate code at each place it is called

> That is not to say we shouldn't consider merging this, but we really need an abstraction to correctly handle this. It is very hard to verify and get right through inspection alone.

There is an element of inception (the movie) in this PR. by copying the output down multiple stages.

perhaps the options are :-
0) Nothing 
1) Limit this to the just the layers not the image/repositories
2) change it to some progress publisher subscribe model, that's going to look a lot like this, except perhaps more verbose since this PR does the publish and handle bit inline.

crosbymichael commented:
@stevvooe @dmcgowan 

Please take a look at this again and make a decision.

stevvooe commented:
@samsabed Part of the issue with this PR is that it assumes a download in a fully asynchronous process when it is still part of a specific http request, which may or may not have completed. The missing aspect here is that we are not treating a download as a first-class resource with a clear lifecycle, making correctness much harder to manage. Effectively, we cannot verify that all poolAdd/Remove calls are correctly made without careful inspection. During future reviews, we have to be mindful that add/remove is correctly add to other places where it is required.

Please at least document the `ProgressStatus` type better. Either prove this is the minimal abstraction or explain each method and how they interact.

That said, this may fix the current issue appropriately.

@crosbymichael I'll defer to @dmcgowan on whether or not we want to merge this. I still think there are some race conditions and the `ProgressStatus` type needs work.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (ps *ProgressStatus) Done() {
+	ps.Lock()
```
Using a lock and a channel isn't very idiomatic. Can we do this without one or the other?

aaronlehmann replied:
I agree with you in principle. I just looked into using a condition variable instead of a channel, but found that a channel was actually better suited to this. A read from a closed channel will finish immediately, but a `Wait()` on a condition variable will always block. That means with a condition variable, we'd need some extra logic to prevent a race between `Wait` and `Done`. So after playing with the alternative, I think this approach is actually very reasonable.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,72 @@
+	ps.Unlock()
+}
+
+func (ps *ProgressStatus) Wait(w io.Writer, msg []byte) error {
```
Is this `Wait` then write to `w`?

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,72 @@
+		logrus.Debugf("Channel is nil ")
+	}
+	if w != nil {
+		w.Write(msg)
```
Please check and report the error.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,72 @@
+	}
+	if w != nil {
+		w.Write(msg)
+		ps.AddObserver(w)
```
This is why I am confused about this method. It says `Wait` on something (whatever channel is), then it writes `msg` to `w`, then adds `w` to the observers?

aaronlehmann replied:
I also found it confusing. I changed it on my branch so that Wait only ways, and the other operations are handled by separate function calls.

tiborvass commented:
@samsabed please rebase

icecrime commented:
@aaronlehmann If this PR is still of interest, would you please be available to carry it (meaning: cherry-pick the author's original commit to preserve his work, but add a commit of yours on top to make it mergeable on master)?

aaronlehmann commented:
I'll take a look.

aaronlehmann commented:
Created a new PR here: #15489

As @icecrime suggested, I cherry-picked @samsabed's original commit, added a merge commit to make it mergeable (which involved significant code movement), and added a few cleanup commits to address @stevvooe's comments and a nit of my own.

Please see that PR for a few open questions I have about this code.

stevvooe commented:
Closing this for #15489.
